app/src/main/java/org/greenstand/android/TreeTracker/managers/Locations.kt

```diff
@@ -0,0 +1,332 @@
+                treeTrackerDAO.insertLocationData(LocationDataEntity(jsonValue))
+            }
+
+            if (isInTreeCaptureMode() && !convergenceWithinRange) {
+
+                var evictedLocation: Location? = if (locationsDeque.size >= CONVERGENCE_DATA_SIZE)
+                    locationsDeque.pollFirst() else null
+                locationsDeque.add(it)
+
+                if (locationsDeque.size >= CONVERGENCE_DATA_SIZE) {
+                    if (convergence == null) {
+                        convergence = Convergence(locationsDeque.toList())
+                        convergence!!.computeConvergence()
+                    } else {
+                        convergence!!.computeSlidingWindowConvergence(evictedLocation!!, it)
+                    }
+                    Timber.d(
+                        "Convergence: Longitude Mean: " +
+                                "[${convergence!!.longitudeConvergence?.mean}]. \n" +
+                                "Longitude standard deviation value: " +
+                                "[${convergence!!.longitudeConvergence?.standardDeviation}]"
+                    )
+                    Timber.d(
+                        "Convergence: Latitude Mean: " +
+                                "[${convergence!!.latitudeConvergence?.mean}]. \n " +
+                                "Latitude standard deviation value: " +
+                                "[${convergence!!.latitudeConvergence?.standardDeviation}]"
+                    )
+                    safeLet(
+                        convergence!!.longitudinalStandardDeviation(),
+                        convergence!!.latitudinalStandardDeviation()
+                    ) { longStdDev, latStdDev ->
+                        convergenceWithinRange = longStdDev < LONG_STD_DEV &&
+                                latStdDev < LAT_STD_DEV
+                    }
+                }
+            }
+        }
+    }
+
+    fun start() {
+        locationUpdateManager.locationUpdateLiveData.observeForever(locationObserver)
+    }
+
+    fun isInTreeCaptureMode(): Boolean {
+        return generatedTreeUuid != null
+    }
+
+    fun turnOnTreeCaptureMode() {
+        generatedTreeUuid = UUID.randomUUID()
+        Timber.d("Convergence: Tree capture mode turned on")
+    }
+
+    fun turnOffTreeCaptureMode() {
+        generatedTreeUuid = null
+        convergence = null
+        locationsDeque.clear()
+        convergenceWithinRange = false
+        Timber.d("Convergence: Tree capture turned off")
+    }
+}
+
+class Convergence(val locations: List<Location>) {
+
+    var longitudeConvergence: ConvergenceStats? = null
+        private set
+    var latitudeConvergence: ConvergenceStats? = null
+        private set
+
+    fun computeStats(data: List<Double>): ConvergenceStats {
+        val mean = data.sum() / data.size
+        var variance = 0.0
+        for (x in data) {
+            variance += (x - mean!!).pow(2.0)
+        }
+        variance /= data.size
```

Comment:
This is fine, but mutable. If we want to avoid that we can do:
`val variance = data.reduce { acc, value -> acc += (value - mean).pow(2.0) } / data.size`

Reply:
This won't work since `acc += ` is not an expression and an assignment. Also, we need every value to be used to calculate the variance and this would miss the very first value as the call would use two values from the data missing the 0th index.

The var here is a local scope, so it is ok. If there is some other nice terse function to do this correctly, will update it.